### ISSUE TITLE
KDE: Invisible Status/Tray Icon - Missing export of share/icons/hicolor/RES/status/?

### ISSUE THREAD

**Issue by Loetkolben** [flathub/me.kozec.syncthingtk#13]
## Bug
When running Syncthing GTK from Flatpak (without the syncthing-gtk dpkg package installed), the status icon in the system tray is invisible: It still exists and "takes up space" in the tray, and interacting with it is still possible (right click opens the context menu), but no icon is visible.

![Blank space in system tray](https://user-images.githubusercontent.com/5988324/63224944-ab6bc780-c1cb-11e9-9510-e81315cb3a83.png)
![Context menu still works](https://user-images.githubusercontent.com/5988324/63224950-b4f52f80-c1cb-11e9-9e9c-2ec4594870e7.png)

## Affected version
at least the following version (I suspect this is also the only one, as support the KDE tray support was just enabled in the last commit: 9bd3403910215b30f99304e50ec8e385f88c2bca)
```
$ LC_ALL=C flatpak info me.kozec.syncthingtk

Syncthing GTK - GUI and notification area icon for Syncthing

          ID: me.kozec.syncthingtk
         Ref: app/me.kozec.syncthingtk/x86_64/stable
        Arch: x86_64
      Branch: stable
     Version: v0.9.4.3
     License: GPL-2.0
      Origin: flathub
  Collection: org.flathub.Stable
Installation: system
   Installed: 65.4?MB
     Runtime: org.gnome.Platform/x86_64/3.32
         Sdk: org.gnome.Sdk/x86_64/3.32

      Commit: da4c0bfb14715341526a54e3775cc44a983f8b9cb6dd3905907677056ed041e9
      Parent: 94cfa16344e9a868df255cf1145c2f7b1aaa7671ac79db33e455faf2c3f40747
     Subject: Enable Status Notifier / System Tray support for KDE (9bd34039)
        Date: 2019-08-08 13:11:30 +0000
```

## Probable cause
I suspect this is caused by not exporting the directories `share/icons/hicolor/RES/status/` to `$XDG_DATA_DIRS` so that KDE is unable to find the icons that have been requested via dbus (shortend log from `dbus-monitor`):
```
method call [...] path=/org/ayatana/NotificationItem/syncthing_gtk; interface=org.freedesktop.DBus.Properties; member=GetAll
   string "org.kde.StatusNotifierItem"
method return [...]
   array [
      dict entry(
         string "Id"
         variant             string "syncthing-gtk"
      [...]
      dict entry(
         string "IconName"
         variant             string "si-syncthing-white-unknown"
      )
     [...]
   ]
```
In for example, `/var/lib/flatpak/app/me.kozec.syncthingtk/current/active/files/share/icons/hicolor/24x24/status/` there exists a `si-syncthing-white-unknown.png` and many other icons - but there are now `status` folders in `/var/lib/flatpak/app/me.kozec.syncthingtk/current/active/export/share/icons/hicolor/RES/`.

Also, the dpkg package (http://download.opensuse.org/repositories/home:/kozec/xUbuntu_18.10/amd64/syncthing-gtk_0.9.4.3_amd64.deb) has an `emblems` folder in `/usr/share/icons/64x64/` - I do not know what role this folder plays in this issue.

If you have the system package for syncthing-gtk installed, it provides the icons in `/usr/share` (at least the packages from http://download.opensuse.org/repositories/home:/kozec/xUbuntu_18.10/ do so) KDE can find the icons and so the problem is not visible.

**Comment by TingPing** [comment]
It should be patched to use an icon name matching the app-id.

**Comment by Loetkolben** [comment]
Well, there are quite many icons from syncthing gtk that need to be renamed:
```
$ ls /var/lib/flatpak/app/me.kozec.syncthingtk/current/active/files/share/icons/hicolor/32x32/status/
si-syncthing-0.png         si-syncthing-black-1.png        si-syncthing-white-0.png
si-syncthing-10.png        si-syncthing-black-2.png        si-syncthing-white-10.png
si-syncthing-11.png        si-syncthing-black-3.png        si-syncthing-white-11.png
si-syncthing-1.png         si-syncthing-black-4.png        si-syncthing-white-1.png
si-syncthing-2.png         si-syncthing-black-5.png        si-syncthing-white-2.png
si-syncthing-3.png         si-syncthing-black-6.png        si-syncthing-white-3.png
si-syncthing-4.png         si-syncthing-black-7.png        si-syncthing-white-4.png
si-syncthing-5.png         si-syncthing-black-8.png        si-syncthing-white-5.png
si-syncthing-6.png         si-syncthing-black-9.png        si-syncthing-white-6.png
si-syncthing-7.png         si-syncthing-black-idle.png     si-syncthing-white-7.png
si-syncthing-8.png         si-syncthing-black-unknown.png  si-syncthing-white-8.png
si-syncthing-9.png         si-syncthing-black-warning.png  si-syncthing-white-9.png
si-syncthing-black-0.png   si-syncthing-idle.png           si-syncthing-white-idle.png
si-syncthing-black-10.png  si-syncthing-unknown.png        si-syncthing-white-unknown.png
si-syncthing-black-11.png  si-syncthing-warning.png        si-syncthing-white-warning.png
```
(Syncthing GTK displays a "spinning" animation when syncthing is currently busy syncing.)

I may be able to look into this further. How should the icons be renamed and how would one go about exporting them from the flatpak app? If you could point me to documentation or examples where this is achieved, that would be nice.

**Comment by TingPing** [comment]
They just need to have the prefix `me.kozec.syncthingtk`, so `me.kozec.syncthingtk.black-0.png` would work for example.

**Comment by Loetkolben** [comment]
Ok, thank you. I'll look into it, when I get the time (kinda busy atm).
Is there any experience on how upstream reacts to these changes / change requests?
Also, sorry for the late response.

**Comment by jurf** [comment]
Upstream is usually quite happy to include things like this, though this might just be easier to patch downstream (see [this](https://github.com/flathub/org.musescore.MuseScore/commit/664f88737623e05422ac9750e34fc6b676843318) for an ugly piece of code doing so).

**Comment by Loetkolben** [comment]
Oh thanks, this looks like what i was looking for :)
I assume flatpak automatically exports any files from `/usr/share` that begin with `APP_NAME`?
If so, i would just need to adapt the snippet you linked into `me.kozec.syncthingtk.yaml` and than modify syncthing-gtk so that it picks up the correct icons? (may take some time though, still busy)

**Comment by TingPing** [comment]
> I assume flatpak automatically exports any files from /usr/share that begin with APP_NAME?

Not quite everything but generally yes.

**Comment by brettwilcox** [comment]
KDE Plasma: 5.23.5
KDE Framework: 5.90
Qt: 5.15.2

Just got the new steam deck running Arch linux and am also having the same issue with the icon.

**Comment by brettwilcox** [comment]
Related issue: https://github.com/kozec/syncthing-gtk/issues/566

```
≻  /usr/bin/flatpak run --branch=stable --arch=x86_64 --command=syncthing-gtk me.kozec.syncthingtk --debug
Gtk-Message: 09:26:53.296: Failed to load module "appmenu-gtk-module"
D UIBuilder     Enabled: header_bar
D UIBuilder     Enabled: is_gnome
D UIBuilder     Enabled: icons_in_menu
D UIBuilder     Loading glade file /app/bin/../share/syncthing-gtk/app.glade
D UIBuilder     Enabled conditions: set([u'icons_in_menu', u'is_gnome', u'header_bar'])
D UIBuilder     Removed 'property' by attribute: is_windows
D UIBuilder     Removed 'property' by attribute: is_windows
D UIBuilder     Allowed node icons_in_menu | old_gtk
D UIBuilder     Removed 'property' by attribute: is_windows
D UIBuilder     Removed node !is_gnome | old_gtk
D UIBuilder     Allowed node icons_in_menu | old_gtk
D UIBuilder     Removed 'property' by attribute: old_gtk
D UIBuilder     Allowed node icons_in_menu | old_gtk
D UIBuilder     Allowed node icons_in_menu | old_gtk
D UIBuilder     Removed 'child' by attribute: traditional_header
D UIBuilder     Allowed node icons_in_menu | old_gtk
D UIBuilder     Removed 'property' by attribute: is_windows
D UIBuilder     Removed node traditional_header
D UIBuilder     Allowed node traditional_header | icons_in_menu
I StatusIcon    Using backend StatusIconGTK3 (primary)
D Daemon        Reading syncthing config /home/plafue/.var/app/me.kozec.syncthingtk/config/syncthing/config.xml
V Notifications Error notifications enabled
V Notifications File update notifications enabled
I App
I App           Syncthing-GTK started and running in notification area
V Daemon        Set refresh interval to 5
W StatusIcon    StatusIcon: Using backend StatusIconAppIndicator (fallback)
E Daemon        g-io-error-quark: Could not connect to 127.0.0.1: Connection refused (39)
Traceback (most recent call last):
  File "/app/lib/python2.7/site-packages/syncthing_gtk/daemon.py", line 1057, in _connected
    self._connection = self.connect_to_service_finish(results)
Error: g-io-error-quark: Could not connect to 127.0.0.1: Connection refused (39)
D App           Creating connect_dialog
[...]
```

https://github.com/flathub/me.kozec.syncthingtk/issues/39 can be merged into this request.

```
pro@Asus:~$ flatpak run me.kozec.syncthingtk
Gtk-Message: 20:39:58.627: Failed to load module "canberra-gtk-module"
I StatusIcon    Using backend StatusIconGTK3 (primary)
Traceback (most recent call last):
  File "/app/lib/python2.7/site-packages/syncthing_gtk/app.py", line 242, in do_command_line
    if self.setup_connection():
  File "/app/lib/python2.7/site-packages/syncthing_gtk/app.py", line 385, in setup_connection
    self.daemon = Daemon()
  File "/app/lib/python2.7/site-packages/syncthing_gtk/daemon.py", line 321, in __init__
    self._read_config()
  File "/app/lib/python2.7/site-packages/syncthing_gtk/daemon.py", line 337, in _read_config
    raise InvalidConfigurationException("Failed to read daemon configuration: %s" % e)
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 26: ordinal not in range(128)
W StatusIcon    StatusIcon: Using backend StatusIconAppIndicator (fallback)

(syncthing-gtk:2): Gdk-CRITICAL **: 20:39:58.852: gdk_window_thaw_toplevel_updates: assertion 'window->update_and_descendants_freeze_count > 0' failed
```

https://github.com/kozec/syncthing-gtk/issues/569 We also have this as a related issue.

```
[](https://github.com/guglovich)Gtk-Message: 14:27:48.065: Failed to load module "canberra-gtk-module"
Gtk-Message: 14:27:48.065: Failed to load module "xapp-gtk3-module"
Gtk-Message: 14:27:48.065: Failed to load module "canberra-gtk-module"
Gtk-Message: 14:27:48.065: Failed to load module "xapp-gtk3-module"
W Configuration Failed to load configuration; Creating new one.
W Configuration Reason: (IOError(2, '\xd0\x9d\xd0\xb5\xd1\x82 \xd1\x82\xd0\xb0\xd0\xba\xd0\xbe\xd0\xb3\xd0\xbe \xd1\x84\xd0\xb0\xd0\xb9\xd0\xbb\xd0\xb0 \xd0\xb8\xd0\xbb\xd0\xb8 \xd0\xba\xd0\xb0\xd1\x82\xd0\xb0\xd0\xbb\xd0\xbe\xd0\xb3\xd0\xb0'),)
I StatusIcon Using backend StatusIconGTK3 (primary)
Traceback (most recent call last):
File "/app/lib/python2.7/site-packages/syncthing_gtk/app.py", line 242, in do_command_line
if self.setup_connection():
File "/app/lib/python2.7/site-packages/syncthing_gtk/app.py", line 385, in setup_connection
self.daemon = Daemon()
File "/app/lib/python2.7/site-packages/syncthing_gtk/daemon.py", line 321, in init
self._read_config()
File "/app/lib/python2.7/site-packages/syncthing_gtk/daemon.py", line 337, in _read_config
raise InvalidConfigurationException("Failed to read daemon configuration: %s" % e)
UnicodeDecodeError: 'ascii' codec can't decode byte 0xd0 in position 10: ordinal not in range(128)
@guglovich
```

**Comment by brettwilcox** [comment]
https://github.com/syncthing/syncthing/commit/01ae866d58bea5823732e739125b5544fd7c87fc is the commit that added/edited the icons to the core Syncthing project.

**Comment by brettwilcox** [comment]
Icons for GTK are generated upstream by 

https://github.com/kozec/syncthing-gtk/blob/master/generate-icons.sh

I see this as a GTK patch for a related issue on another project. It may be useful here, but I'm not a GTK dev.

https://github.com/mydongistiny/paper-icon-theme-git/commit/227fe0a48571f7898262e4496291e3d59f364e10

**Comment by muelli** [comment]
the path forward to fixing this up has been [outlined before](https://github.com/flathub/me.kozec.syncthingtk/issues/13#issuecomment-522901989), no?
Somebody needs to step up to rename all those icons to match the APP_ID.

**Comment by brettwilcox** [comment]
> the path forward to fixing this up has been [outlined before](https://github.com/flathub/me.kozec.syncthingtk/issues/13#issuecomment-522901989), no? Somebody needs to step up to rename all those icons to match the APP_ID.

[jurf](https://github.com/flathub/me.kozec.syncthingtk/issues/13#issuecomment-546967577) provided a link to code that we can modify for our needs.

We need to change the [syncthing.yaml](https://github.com/flathub/me.kozec.syncthingtk/blob/master/syncthing.yaml) to copy the icons into the folder and append the app ID during the copy.

Here is pseudo code that should get us closer.

```
    post-install:
      # Add prefixes to mimetype icons so they can be exported
      # TODO: really, this bit could be nicer
      - cd $XDG_DATA_HOME/icons/hicolor;
        for d in */status/; do
          for f in ${d}*; do
            mv "$f" "${d}me.kozec.syncthingtk.$(basename $f)";
          done;
        done
```

**Comment by TingPing** [comment]
You can just move files, the application has to reference the new name. This all belongs upstream.

**Comment by HugLifeTiZ** [comment]
> I'm trying out KDE Neon on a new computer and am running into this as well. I thought I could get around it by copying all the icons from upstream into `$XDG_DATA_HOME/icons/hicolor`, but that didn't work.

So this thing that I mentioned didn't work suddenly started working at some point. I don't remember if I did anything special between when it wasn't working and when it was. But this should serve as a workaround for now.

**Comment by TingPing** [comment]
The code can be patched here but its disappointing.

**Comment by brettwilcox** [comment]
@TingPing Yeah, I agree. Let me reach out on the original issue before any code is produced here and see if the original owner has had a change of heart. With the new Steam Deck hitting the market, there are a lot of new users trying out Linux (Arch specifically) for the first time. 

What you have done with Flatpak is amazing! I want everyone to have an excellent experience no matter the app.

**Comment by brettwilcox** [comment]
Proposed upstream fix. Waiting on owner for response.

https://github.com/kozec/syncthing-gtk/issues/566#issuecomment-1107633100

**Comment by TingPing** [comment]
Go for it.

**Comment by CobraA1** [comment]
Any update on this issue? It's been a few months . . .

**Comment by lsellens** [comment]
Looks like the original author has abandoned it. Why not switch to a more active fork such as https://github.com/syncthing-gtk/syncthing-gtk

**Comment by muelli** [comment]
ah, yeah, looks interesting! Thanks for the pointer. Is that a fork?
Let me look into providing a flatpak for this app.

**Comment by muelli** [comment]
ah, no, wait. 
We have been using https://github.com/syncthing-gtk/syncthing-gtk for ever.
Let me try to investigate why the tag https://github.com/syncthing-gtk/syncthing-gtk/commit/c001b6679c956ddaa967c2d61f96504245818f5c didn't cause the bot to pick it up.

https://release-monitoring.org/project/17342/ is configured to follow https://github.com/syncthing/syncthing-gtk/tags which makes sense. But somehow Github redirects it to https://github.com/kozec/syncthing-gtk/tags. And there, the new tag doesn't show up.

It seem https://github.com/syncthing/syncthing-gtk/ is redirecting to https://github.com/kozec/syncthing-gtk. That seems weird, given that development seems to happen in the former, not the latter. I don't know whose fault it is. Maybe @kozec or @Rudd-O can ensure that the redirect doesn't happen.

**Comment by lsellens** [comment]
> It seem https://github.com/syncthing/syncthing-gtk/ is redirecting to https://github.com/kozec/syncthing-gtk.

url should be https://github.com/syncthing-gtk/syncthing-gtk not https://github.com/syncthing/syncthing-gtk

**Comment by TingPing** [comment]
> sounds like release-monitoring.org needs some convincing. Whoever feels responsible for syncthing-gtk should try to get the tracking URL changed.

Anybody can update it. I've updated it now.

**Comment by TingPing** [comment]
It seems slightly confused by tags like `Syncthing_v1.10.0` existing being newer than `v0.9.4.4.2`

**Comment by ojkastl** [comment]
Any update on this? Still having this issue on Fedora 39.

(The flatpak is still called `me.kozec.syncthingtk` even if it uses https://github.com/syncthing-gtk/syncthing-gtk, but that is another story)

**Comment by muelli** [comment]
If you feel so inclined you can try to make something like this work:
https://github.com/flathub/me.kozec.syncthingtk/issues/13#issuecomment-1103778927

**Comment by lsellens** [comment]
Oddly enough this started working on the steam deck a while ago. The icon is low resolution and likes like crap but it does show up.
